streampipes-data-explorer/src/main/java/org/apache/streampipes/dataexplorer/v4/query/StreamedQueryResultProvider.java

```diff
@@ -0,0 +1,113 @@
+import static org.apache.streampipes.dataexplorer.v4.SupportedDataLakeQueryParameters.QP_PAGE;
+
+public class StreamedQueryResultProvider extends QueryResultProvider {
+
+  private static final int MAX_RESULTS_PER_QUERY = 500000;
+
+  private final OutputFormat format;
+
+  public StreamedQueryResultProvider(ProvidedQueryParams params,
+                                     OutputFormat format,
+                                     boolean ignoreMissingValues) {
+    super(params, ignoreMissingValues);
+    this.format = format;
+  }
+
+  public void getDataAsStream(OutputStream outputStream) throws IOException {
+    var usesLimit = queryParams.has(QP_LIMIT);
+    var configuredWriter = ConfiguredOutputWriter
+        .getConfiguredWriter(format, queryParams, ignoreMissingData);
+
+    if (!queryParams.has(QP_LIMIT)) {
+      queryParams.update(QP_LIMIT, MAX_RESULTS_PER_QUERY);
+    }
+
+    var limit = queryParams.getAsInt(QP_LIMIT);
+    var measurement = findByMeasurementName(queryParams.getMeasurementId()).get();
+
+    SpQueryResult dataResult;
+    int page = 0;
+    if (queryParams.has(QP_PAGE)) {
+      page = queryParams.getAsInt(QP_PAGE);
+    }
+
+    boolean isFirstDataItem = true;
+    configuredWriter.beforeFirstItem(outputStream);
+    do {
+      queryParams.update(SupportedDataLakeQueryParameters.QP_PAGE, String.valueOf(page));
+      dataResult = getData();
+
+      if (dataResult.getTotal() > 0) {
+        changeTimestampHeader(measurement, dataResult);
+        var columns = dataResult.getHeaders();
+        for (List<Object> row : dataResult.getAllDataSeries().get(0).getRows()) {
+          configuredWriter.writeItem(outputStream, row, columns, isFirstDataItem);
+          isFirstDataItem = false;
+        }
+      }
+      page++;
+    } while (queryNextPage(dataResult.getTotal(), usesLimit, limit, page));
+    configuredWriter.afterLastItem(outputStream);
+  }
+
+  private boolean queryNextPage(int lastResultsCount,
+                                boolean usesLimit,
+                                int limit,
+                                int lastPage) {
+    if (usesLimit) {
+      return !(limit <= (lastPage) * MAX_RESULTS_PER_QUERY);
+    } else {
+      return lastResultsCount > 0;
+    }
+  }
+
+  private Optional<DataLakeMeasure> findByMeasurementName(String measurementName) {
+    return DataExplorerUtils.getInfos()
+        .stream()
+        .filter(measurement -> measurement.getMeasureName().equals(measurementName))
+        .findFirst();
+  }
+
+  /**
+   * Replaces the field 'time' of the data result with the actual timestamp field name of the measurement
+   * @param measurement contains the actual timestamp name value
+   * @param dataResult the query result of the database with 'time' as timestamp field name
+   */
+  private void changeTimestampHeader(DataLakeMeasure measurement, SpQueryResult dataResult) {
+    dataResult.getHeaders().set(0, measurement.getTimestampFieldName());
```

Comment:
Can we rely on `time` being always the first header?

Reply:
Thanks, fixed!